src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs

```diff
@@ -5004,4 +5004,84 @@ partial class C : I<FI>
         var verifier = CompileAndVerify(new[] { (source0, "F0.cs"), (source1, "F1.cs"), (source2, "F2.cs") }, expectedOutput: "12");
         verifier.VerifyDiagnostics();
     }
+
+    [Fact]
+    public void ShadowNamespace_01()
+    {
+        var source1 = """
+            namespace App.Widget
+            {
+                class Inner { }
+            }
+
+            """;
+
+        var source2 = """
+            namespace App
+            {
+                file class Widget { }
+            }
+
+            """;
+
+        var comp = CreateCompilation(new[] { (source1, "File1.cs"), (source2, "File2.cs") });
+        comp.VerifyDiagnostics();
+
+        comp = CreateCompilation(source1 + source2);
+        comp.VerifyDiagnostics(
+            // (7,16): error CS9071: The namespace 'App' already contains a definition for 'Widget' in this file.
+            //     file class Widget { }
+            Diagnostic(ErrorCode.ERR_FileLocalDuplicateNameInNS, "Widget").WithArguments("Widget", "App").WithLocation(7, 16));
+
+        comp = CreateCompilation(source2 + source1);
+        comp.VerifyDiagnostics(
+            // (3,16): error CS9071: The namespace 'App' already contains a definition for 'Widget' in this file.
+            //     file class Widget { }
+            Diagnostic(ErrorCode.ERR_FileLocalDuplicateNameInNS, "Widget").WithArguments("Widget", "App").WithLocation(3, 16));
+    }
```

Comment:
>} [](http://example.com/codeflow?start=4&length=1)

Consider also testing these same files, but where source1 is in a separate compilation: both as a compilation reference and as a metadata reference.